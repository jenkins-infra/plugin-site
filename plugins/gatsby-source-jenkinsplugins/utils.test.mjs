/**
 * @jest-environment node
 */
import fs from 'fs';
import path, {dirname} from 'path';
import {
    fetchPluginData,
    fetchPluginHealthScore,
    fixGitHubUrl,
    markdownToHtml,
} from './utils.mjs';
import nock from 'nock';
import {jest, describe, beforeEach, it, expect, afterEach} from '@jest/globals';

import {fileURLToPath} from 'url';
import {createContentDigest} from 'gatsby-core-utils';

const __dirname = dirname(fileURLToPath(import.meta.url));
const readText = async (fileName) => {const buffer = await fs.promises.readFile(path.join(__dirname, '__mocks__', fileName)); return buffer.toString();};

describe('utils', () => {
    let _reporter;
    beforeEach(async () => {
        process.env.GET_CONTENT = true;
        nock.disableNetConnect();

        _reporter = {
            panic: (...args) => {throw args[0];},
            error: jest.fn(),
            activityTimer: () => {
                return {
                    start: jest.fn(),
                    end: jest.fn(),
                };
            }
        };
    });
    afterEach(() => {
        nock.cleanAll();
    });
    it('fix GitHub URL: submodule gets expanded', () => {
        expect(fixGitHubUrl('https://github.com/jenkinsci/blueocean-plugin/blueocean-bitbucket-pipeline', 'master'))
            .toBe('https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-bitbucket-pipeline');
    });
    it('fix GitHub URL: expanded stays expanded', () => {
        expect(fixGitHubUrl('https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-bitbucket-pipeline', 'master'))
            .toBe('https://github.com/jenkinsci/blueocean-plugin/tree/master/blueocean-bitbucket-pipeline');
    });
    it('fix GitHub URL: no submodule, keep short', () => {
        expect(fixGitHubUrl('https://github.com/jenkinsci/junit-plugin', ''))
            .toBe('https://github.com/jenkinsci/junit-plugin');
    });
    it('get plugin data for a wiki based plugin', async () => {
        nock('https://updates.jenkins.io')
            .get('/update-center.actual.json')
            .reply(200, JSON.parse(await readText('update-center.actual.json')));
        nock('https://updates.jenkins.io')
            .get('/current/plugin-documentation-urls.json')
            .reply(200, '{"ios-device-connector":{"url":"https://wiki.jenkins-ci.org/display/JENKINS/iOS+Device+Connector+Plugin"}}');
        nock('https://www.jenkins.io')
            .get('/doc/pipeline/steps/contents.json')
            .reply(200, []);
        nock('https://www.jenkins.io')
            .get('/doc/developer/extensions/contents.json')
            .reply(200, []);
        nock('https://raw.githubusercontent.com:443')
            .get('/jenkinsci/bom/master/bom-weekly/pom.xml')
            .reply(200, '');
        nock('https://raw.githubusercontent.com:443')
            .get('/jenkinsci/jenkins/master/core/src/main/resources/jenkins/split-plugins.txt')
            .reply(200, await readText('split-plugins.txt'));
        nock('https://raw.githubusercontent.com:443')
            .get('/jenkinsci/jenkins/master/core/src/main/resources/jenkins/canonical-labels.txt')
            .reply(200, '');
        nock('https://plugins.jenkins.io').persist()
            .get(/\/api\/plugin\/.*/)
            .reply(200, '{"wiki":{"content": "<p>This plugin lists up all the iOS devices connected to the master and all the Jenkins slaves, and provide operations to them.</p>"}}');

        const createNode = jest.fn().mockResolvedValue();
        const createNodeId = jest.fn(key => key);
        const firstReleases = {'ios-device-connector': new Date(0)};
        const labelToCategory = {'ios': 'languagesPlatforms', 'builder': 'buildManagement'};
        const stats = {
            'core': {'installations': [{total: 300000}, {total: 300000}]},
            'ios-device-connector': {
                'currentInstalls': 269, 'installations': [
                    {'timestamp': 1590984000000, 'total': 349},
                    {'timestamp': 1593576000000, 'total': 348},
                    {'timestamp': 1596254400000, 'total': 341},
                    {'timestamp': 1598932800000, 'total': 347},
                    {'timestamp': 1601524800000, 'total': 318},
                    {'timestamp': 1604203200000, 'total': 325},
                    {'timestamp': 1606798800000, 'total': 312},
                    {'timestamp': 1609477200000, 'total': 301},
                    {'timestamp': 1612155600000, 'total': 297},
                    {'timestamp': 1614574800000, 'total': 303},
                    {'timestamp': 1617249600000, 'total': 290},
                    {'timestamp': 1619841600000, 'total': 275},
                    {'timestamp': 1622520000000, 'total': 269}]
            }
        };
        await fetchPluginData({createNode, createNodeId, createContentDigest, reporter: _reporter, firstReleases, labelToCategory, stats});
        expect(createNode.mock.calls.filter(call => call[0].name === 'ios-device-connector').map(args => args[0])).toMatchSnapshot();
    }, 15000);
    it('get plugin healthScore data', async () => {
        nock('https://plugin-health.jenkins.io')
            .get('/api/scores')
            .reply(200, JSON.parse(await readText('plugin-health-score.json')));
        const createNode = jest.fn().mockResolvedValue();

        await fetchPluginHealthScore({createNode, reporter: _reporter});
        expect(createNode.mock.calls.filter(call => call[0].id === 'aws-java-sdk-sns').map(args => args[0])).toMatchSnapshot();
    });
    it ('should convert markdown to html', async () => {
        const translated = await markdownToHtml(await readText('plugin-doc.md'));
        expect(translated).toMatchSnapshot();
    });
    it ('should convert markdown with tables to html', async () => {
        const translated = await markdownToHtml(await readText('table.md'));
        expect(translated).toMatchSnapshot();
    });
});
