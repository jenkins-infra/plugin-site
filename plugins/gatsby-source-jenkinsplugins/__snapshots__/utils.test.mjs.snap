// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils get plugin data for a wiki based plugin 1`] = `
[
  {
    "baseHref": "https://wiki.jenkins-ci.org/display/JENKINS/",
    "id": "ios-device-connector <<< JenkinsPluginWiki",
    "internal": {
      "content": "MISSING",
      "contentDigest": "10d1208b485425756fcc932229386b02",
      "mediaType": "text/pluginhtml",
      "type": "JenkinsPluginWiki",
    },
    "name": "ios-device-connector",
    "url": "https://wiki.jenkins-ci.org/display/JENKINS/iOS+Device+Connector+Plugin",
  },
  {
    "buildDate": 2012-10-10T00:00:00.000Z,
    "children": [],
    "defaultBranch": "master",
    "dependencies": [
      {
        "implied": true,
        "name": "external-monitor-job",
        "optional": false,
        "title": "External Monitor Job Type",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "ldap",
        "optional": false,
        "title": "LDAP",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "pam-auth",
        "optional": false,
        "title": "PAM Authentication",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "mailer",
        "optional": false,
        "title": "Mailer",
        "version": "1.2",
      },
      {
        "implied": true,
        "name": "matrix-auth",
        "optional": false,
        "title": "Matrix Authorization Strategy",
        "version": "1.0.2",
      },
      {
        "implied": true,
        "name": "windows-slaves",
        "optional": false,
        "title": "WMI Windows Agents",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "antisamy-markup-formatter",
        "optional": false,
        "title": "OWASP Markup Formatter",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "matrix-project",
        "optional": false,
        "title": "Matrix Project",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "junit",
        "optional": false,
        "title": "JUnit",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "bouncycastle-api",
        "optional": false,
        "title": "bouncycastle API",
        "version": "2.16.0",
      },
      {
        "implied": true,
        "name": "command-launcher",
        "optional": false,
        "title": "Command Agent Launcher",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "jdk-tool",
        "optional": false,
        "title": "Oracle Java SE Development Kit Installer",
        "version": "1.0",
      },
      {
        "implied": true,
        "name": "jaxb",
        "optional": false,
        "title": "JAXB",
        "version": "2.3.0",
      },
      {
        "implied": true,
        "name": "trilead-api",
        "optional": false,
        "title": "Trilead API",
        "version": "1.0.4",
      },
      {
        "implied": true,
        "name": "sshd",
        "optional": false,
        "title": "SSHD",
        "version": "3.0.1",
      },
    ],
    "developers": [],
    "excerpt": "Talks to iOS devices connected to slaves and do stuff.",
    "firstRelease": "1970-01-01T00:00:00.000Z",
    "gav": "org.jenkins-ci.plugins:ios-device-connector:1.2",
    "hasBomEntry": false,
    "hasExtensions": false,
    "hasPipelineSteps": false,
    "id": "ios-device-connector <<< JenkinsPlugin",
    "internal": {
      "contentDigest": "1da1c7c8d56f5351f4d3c618759f7689",
      "type": "JenkinsPlugin",
    },
    "issueTrackers": [
      {
        "reportUrl": "https://www.jenkins.io/participate/report-issue/redirect/#17020",
        "type": "jira",
        "viewUrl": "https://issues.jenkins.io/issues/?jql=component=17020",
      },
    ],
    "labels": [
      "builder",
      "ios",
    ],
    "name": "ios-device-connector",
    "parent": null,
    "popularity": 275,
    "previousTimestamp": "2012-10-08T20:04:36.00Z",
    "previousVersion": "1.1",
    "releaseTimestamp": "2012-10-10T17:27:42.00Z",
    "requiredCore": "1.466",
    "scm": "https://github.com/jenkinsci/ios-device-connector-plugin",
    "securityWarnings": [],
    "sha1": "B2z2wTyA8l5uwz9HqD+xhG9Hjns=",
    "sha256": "Tray3+tyfE6xoTu1Rge9+INkcj5y7Bgw7nPm//saiYw=",
    "size": 1548862,
    "stats": {
      "currentInstalls": 269,
      "installations": [
        {
          "timestamp": 1590984000000,
          "total": 349,
        },
        {
          "timestamp": 1593576000000,
          "total": 348,
        },
        {
          "timestamp": 1596254400000,
          "total": 341,
        },
        {
          "timestamp": 1598932800000,
          "total": 347,
        },
        {
          "timestamp": 1601524800000,
          "total": 318,
        },
        {
          "timestamp": 1604203200000,
          "total": 325,
        },
        {
          "timestamp": 1606798800000,
          "total": 312,
        },
        {
          "timestamp": 1609477200000,
          "total": 301,
        },
        {
          "timestamp": 1612155600000,
          "total": 297,
        },
        {
          "timestamp": 1614574800000,
          "total": 303,
        },
        {
          "timestamp": 1617249600000,
          "total": 290,
        },
        {
          "timestamp": 1619841600000,
          "total": 275,
        },
        {
          "timestamp": 1622520000000,
          "total": 269,
        },
      ],
      "trend": -20,
    },
    "title": "iOS Device connector",
    "url": "https://updates.jenkins.io/download/plugins/ios-device-connector/1.2/ios-device-connector.hpi",
    "version": "1.2",
  },
]
`;

exports[`utils get plugin healthScore data 1`] = `
[
  {
    "children": [],
    "details": [
      {
        "components": [
          {
            "reasons": [
              "Plugin does not seem to have on-going security advisory.",
            ],
            "value": 100,
            "weight": 1,
          },
        ],
        "name": "security",
        "value": 100,
        "weight": 1,
      },
      {
        "components": [
          {
            "reasons": [
              "There is less than 6 months between the last release and the last commit.",
            ],
            "value": 100,
            "weight": 1,
          },
          {
            "reasons": [
              "There is less than 4 years between the last commit and the last release.",
            ],
            "value": 100,
            "weight": 1,
          },
          {
            "reasons": [
              "No commit in the last 4 years.",
            ],
            "value": 0,
            "weight": 0,
          },
          {
            "reasons": [
              "There is less than a year between the last commit and the last release.",
            ],
            "value": 100,
            "weight": 1,
          },
          {
            "reasons": [
              "The plugin is not marked as up for adoption",
            ],
            "value": 100,
            "weight": 1,
          },
          {
            "reasons": [
              "There is less than 2 years between the last commit and the last release.",
            ],
            "value": 100,
            "weight": 1,
          },
        ],
        "name": "adoption",
        "value": 100,
        "weight": 0.8,
      },
      {
        "components": [
          {
            "reasons": [
              "The plugin appears in the update-center.",
            ],
            "value": 100,
            "weight": 1,
          },
        ],
        "name": "update-center-plugin-publication",
        "value": 100,
        "weight": 1,
      },
      {
        "components": [
          {
            "reasons": [
              "Documentation is in plugin repository.",
            ],
            "value": 100,
            "weight": 15,
          },
          {
            "reasons": [
              "JEP-229 is configured on the plugin.",
            ],
            "value": 100,
            "weight": 5,
          },
          {
            "reasons": [
              "Jenkinsfile detected in plugin repository.",
            ],
            "value": 100,
            "weight": 65,
          },
          {
            "reasons": [
              "dependabot is configured.",
              "3 open dependency pull requests",
            ],
            "value": 0,
            "weight": 15,
          },
        ],
        "name": "repository-configuration",
        "value": 85,
        "weight": 0.5,
      },
      {
        "components": [
          {
            "reasons": [
              "Plugin is not marked as deprecated.",
            ],
            "value": 100,
            "weight": 1,
          },
        ],
        "name": "deprecation",
        "value": 100,
        "weight": 0.8,
      },
    ],
    "id": "aws-java-sdk-sns",
    "internal": {
      "contentDigest": "f789c60332b91b320525aea3639a7c73",
      "type": "JenkinsPluginHealthScore",
    },
    "parent": null,
    "value": 98,
  },
]
`;

exports[`utils should convert markdown to html 1`] = `
"<p>This plugin allows user to configure additional identities in jenkins
user database for external services, typically SCM repositories.</p>
<h2>Introduction</h2>
<p>As SCM plugin parse changelog, they have to map SCM committers IDs with
jenkins user database. This results in many cases in user duplication,
until you exactly have the same ID in jenkins and SCM.</p>
<p>Jenkins 1.480 introduces an extension point to resolve jenkins user
"canonical" ID when searching for user in Database by id or full name.
This plugin uses this extension point to let user configure external
identities as user properties.</p>
<h2>Usage</h2>
<p>On my Jenkins instance, I'm authenticated as "nicolas". As I want to use
the same identity for commits on repositories, I can setup an additional
identity for my account on googlecode :
<img src="docs/images/Capture_d%E2%80%99%C3%A9cran_2012-08-21_%C3%A0_15.03.47.png" alt=""></p>
<p>With this additional identity set, Jenkins will be able to match the
committer id in svn "<a href="mailto:nicolas.deloof@gmail.com">nicolas.deloof@gmail.com</a>" with the jenkins user
"nicolas", and link the builds I contributed to in my user view :</p>
<p><img src="docs/images/Capture_d%E2%80%99%C3%A9cran_2012-08-21_%C3%A0_15.08.01.png" alt=""></p>
<h2>Realms</h2>
<p>As for HTTP authentication, a realm can be set to restrict an identity
to a set of network resources (i.e. domain names in most cases). The
<em>realm</em> attribute can be used to restrict the sources user ID matching
will apply. In most cases, this is a substring of the SCM repository
URL. If not set, additional identity applies to all user lookups,
whatever the id source is.</p>
<p>This feature requires SCM plugins to be updated so that they compute
host information form changelog and pass it to extension as
<a href="http://javadoc.jenkins-ci.org/hudson/model/User.CanonicalIdResolver.html#REALM">REALM</a> context
attribute. Those plugins have been updated to support this advanced
feature :</p>
<ul>
<li>TO BE COMPLETED</li>
</ul>
<h2>Tips</h2>
<p>git plugin uses user name, as set in git commit, as committer ID, so you
don't need an additional identity, just ensure your git client is
configured with correct user name set :</p>
<pre><code class="language-syntaxhighlighter-pre">git config --global user.name "Your Full Name Comes Here"
</code></pre>
<h2>Changelog</h2>
<h3>1.1 (Oct 20 2015)</h3>
<ul>
<li><a href="https://issues.jenkins-ci.org/browse/JENKINS-28181">JENKINS-28181</a>
NPE thrown in certain cases.</li>
<li>Internal class rename.</li>
<li>Missing descriptor error.</li>
</ul>
<h3>1.0</h3>
<ul>
<li>initial release</li>
</ul>"
`;

exports[`utils should convert markdown with tables to html 1`] = `
"<table>
<thead>
<tr>
<th>Month</th>
<th>Savings</th>
</tr>
</thead>
<tbody>
<tr>
<td>January</td>
<td>$250</td>
</tr>
<tr>
<td>February</td>
<td>$80</td>
</tr>
<tr>
<td>March</td>
<td>$420</td>
</tr>
</tbody>
</table>"
`;
